preCICE does not build with Boost 1.65.1 and Boost 1.66

From commit b216b84aeda6387670fe5129f5e1a3280f888f81, preCICE does not build successfully when using Ubuntu 18.04. Boost is installed using `apt`. Before b216b84aeda6387670fe5129f5e1a3280f888f81 building works fine. 

I think that we should still try to fix this, since it is a nice feature to be able to install all preCICE dependencies using `apt`, if you are using Ubuntu 18.04.

# Working 

* Ubuntu 18.04 with boost 1.67 from source (https://github.com/precice/precice/issues/176#issuecomment-434742389)
* Ubuntu 18.04 with boost 1.60 from source (https://github.com/precice/precice/issues/176#issuecomment-434758016)
* Ubuntu 16.04 with boost 1.67 from source (https://github.com/precice/precice/issues/176#issuecomment-434741395)
* Ubuntu 14.04 with boost 1.60 from source (https://github.com/precice/precice/issues/176#issuecomment-434741395)

# Not Working

* Ubuntu 18.04 with boost 1.65.1 from apt-get (https://github.com/precice/precice/issues/176#issuecomment-434730078); [Dockerfile](https://github.com/precice/systemtests/blob/ubuntu1804base/Dockerfile.precice1804)
* Ubuntu 18.04 with boost 1.65.1 from source (https://github.com/precice/precice/issues/176#issuecomment-434766867)
* Ubuntu 16.04 with boost 1.65.1 from source (https://github.com/precice/precice/issues/176#issuecomment-434790126)
* Ubuntu 16.04 with boost 1.66 from source (https://github.com/precice/precice/issues/176#issuecomment-434792138)

*For sake of completeness:*

If I run the following command
```
docker build -t precice1804boost1651 -f Dockerfile.precice1804boost1651 .
```
I get the following compilation error:
```
...
mpicxx -o build/debug-nopetsc-nopython/com/tests/SocketCommunicationTest.o -c -Wall -Wextra -Wno-unused-parameter -std=c++11 -fPIC -g3 -O0 -DPRECICE_NO_PETSC -DBOOST_ALL_DYN_LINK -DBOOST_ASIO_ENABLE_OLD_SERVICES -DMPICH_SKIP_MPICXX -DPRECICE_NO_PYTHON -Isrc -I/usr/include/eigen3 -I/usr/include/libxml2 src/com/tests/SocketCommunicationTest.cpp
In file included from /usr/local/include/boost/test/tools/collection_comparison_op.hpp:18:0,
                 from /usr/local/include/boost/test/test_tools.hpp:55,
                 from /usr/local/include/boost/test/unit_test.hpp:18,
                 from src/testing/Testing.hpp:5,
                 from src/com/tests/CommunicateMeshTest.cpp:10:
/usr/local/include/boost/test/utils/is_forward_iterable.hpp: In instantiation of 'struct boost::unit_test::bt_iterator_traits<precice::mesh::Triangle, true>':
/usr/local/include/boost/test/tools/collection_comparison_op.hpp:437:1:   required from 'struct boost::test_tools::assertion::op::EQ<precice::mesh::Triangle, precice::mesh::Triangle, void>'
/usr/local/include/boost/test/tools/assertion.hpp:326:7:   required from 'class boost::test_tools::assertion::binary_expr<boost::test_tools::assertion::value_expr<precice::mesh::Triangle&>, precice::mesh::Triangle&, boost::test_tools::assertion::op::EQ<precice::mesh::Triangle, precice::mesh::Triangle, void> >'
src/com/tests/CommunicateMeshTest.cpp:125:7:   required from here
/usr/local/include/boost/test/utils/is_forward_iterable.hpp:206:36: error: no type named 'value_type' in 'class precice::mesh::Triangle'
     typedef typename T::value_type value_type;
                                    ^~~~~~~~~~
scons: *** [build/debug-nopetsc-nopython/com/tests/CommunicateMeshTest.o] Error 1
scons: building terminated because of errors.
The command '/bin/sh -c scons petsc=$petsc_para mpi=$mpi_para python=$python_para -j$(nproc)' returned a non-zero code: 2
```

--- a/src/mesh/Quad.hpp
+++ b/src/mesh/Quad.hpp
@@ -27,6 +27,9 @@ public:
   /// Type of the random access vertex iterator
   using iterator = const_iterator; //IndexRangeIterator<Quad, Eigen::Vector3d>;
 
+  /// Fix for the Boost.Test versions 1.65 - 1.67.1
+  using value_type = Eigen::VectorXd;
+
   /// Constructor, the order of edges defines the outer normal direction.
   Quad(
       Edge &edgeOne,

--- a/src/mesh/Triangle.hpp
+++ b/src/mesh/Triangle.hpp
@@ -41,6 +41,8 @@ public:
   /// Type of the read-only random access vertex iterator
   using iterator = const_iterator;
 
+  /// Fix for the Boost.Test versions 1.65.1 - 1.67
+  using value_type = Eigen::VectorXd;
 
   /// Constructor, the order of edges defines the outer normal direction.
   Triangle(

